src/python/pants/backend/codegen/protobuf/target_types.py

```diff
@@ -27,3 +45,74 @@ class ProtobufLibrary(Target):
     alias = "protobuf_library"
     core_fields = (*COMMON_TARGET_FIELDS, ProtobufDependencies, ProtobufSources, ProtobufGrpcToggle)
     help = f"Protobuf files used to generate various languages.\n\nSee f{doc_url('protobuf')}."
+
+
+logger = logging.getLogger(__name__)
+
+
+_T = TypeVar("_T", bound=Target)
+
+
+@union
+@dataclass(frozen=True)
+class GenerateTargetsRequest:
+    target_class: ClassVar[type[_T]]
+    target: _T
+
+
+class GeneratedTargets(Collection[Target]):
+    pass
+
+
+class GenerateProtobufLibraryFromProtobufLibrary(GenerateTargetsRequest):
+    target_class = ProtobufLibrary
+
+
+@rule
+async def generate_protobuf_library_from_protobuf_library(
+    request: GenerateProtobufLibraryFromProtobufLibrary, union_membership: UnionMembership
+) -> GeneratedTargets:
+    paths = await Get(SourcesPaths, SourcesPathsRequest(request.target[ProtobufSources]))
+    return GeneratedTargets(
+        generate_file_level_target(ProtobufLibrary, request.target, union_membership, file_path=fp)
```

Comment:
NB: this is not called `file_target`. Per https://docs.google.com/document/d/1HpJn2jTWf5sKob6zhe4SqHZ7KWBlX4a6OJOwnuw-IHo/edit, it's instead going to be a "normal target" - only that the metadata describes a single file. `python_source` and `python_test` target, rather than `python_sources`.